### PR TITLE
fix shm deserialize data error for those data with no fixed size

### PR DESCRIPTION
**Pull request by SeanYu81**
link to the issue: https://github.com/eclipse-cyclonedds/cyclonedds-cxx/issues/407

### CODE REVIEW

**Comment by eboasson** [comment]
Thank you very much @SeanYu81!

After looking at the proposed change and at the CI failures, it looks like you fixed the case where the data gets transferred through shared memory in serialized form (for example because it contains a `std::string`, and we can't copy that from one process into the next) but broke the case where the data does not need to be serialized (for example a type with only some integers).

I think both the old and the new code need to be present, choosing the one based on what is actually in memory. If you're happy to take a look at that, then that would be great 😊 

The other obvious thing we are missing is a test case for the case that failed ... but that's not a blocker for merging a fix, just something that really must be done soon.

**Comment by SeanYu81** [comment]
Hello @eboasson,

  Thanks for your comments.
  You are right, and I make a mistake to ignore logic for fixed size data. I will make changes again and please help to review.

Regards,
Yu, Zhi Qiang

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp [hunk 1]
```diff
@@ -647,9 +647,21 @@ ddsi_serdata * serdata_from_iox_buffer(
     }
 
     // key handling
-    const auto& msg = *static_cast<const T*>(d->iox_chunk);
-    d->key_md5_hashed() = to_key(msg, d->key());
-    d->populate_hash();
+    if (typecmn->fixed_size == 1) {
```
Comment (eboasson) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:650]: Nitpick:
```suggestion
    if (typecmn->fixed_size) {
```
(or, alternatively, `typecmn->fixed_size != 0`)

What you wrote is perfectly correct given that `fixed_size` is a single bit, but it seems a possible risk if anything gets refactored.

**Review by eboasson — COMMENTED** [review]
Thanks @SeanYu81 this looks good to me. I'd have merged it if I didn't have a nitpick-y comment. Would you please have a look at that?

**Comment by eboasson** [comment]
There's one more thing, @SeanYu81: the Eclipse Foundation requires that the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php) is signed (the "eclipsefdn/eca" check). Would you be willing to sign that?

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp (hunk 1 above)
Comment (SeanYu81) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:650]: hi @eboasson Your change is great and it will make codes more robust. Thanks~

**Comment by SeanYu81** [comment]
> There's one more thing, @SeanYu81: the Eclipse Foundation requires that the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php) is signed (the "eclipsefdn/eca" check). Would you be willing to sign that?

hi @eboasson No problem, and I have signed the ECA.

**Pull request merged**